Deploying to Android fails with zipalign error

_From @johnshardman on April 4, 2018 10:25_

Bug report best practices: https://github.com/xamarin/Xamarin.Forms/wiki/Submitting-Issues

### Description
After upgrading Visual Studio to 15.6, deployment of app to Android (physical devices and emulators) failed with a zipalign error.

### Steps to Reproduce

1. Check the "Generate one package (.apk) per selected ABI" option in the Android project properties.
2. Hit F5

### Expected Behavior

App should build, deploy and run in debug mode.

### Actual Behavior

Deployment error reported:

AndroidZipAlign Task
Alignment: 4
Source: G:\tfs\MyApp\Top\MyApp.Android\obj\Debug\android\bin\com.MyApp.MyApp-arm64-v8a.apk
DestinationDirectory: bin\Debug\
C:\ProgramFiles\Android\AndroidSDK\build-tools\25.0.2\zipalign.exe 4 "G:\tfs\MyApp\Top\MyApp.Android\obj\Debug\android\bin\com.MyApp.MyApp-arm64-v8a.apk" "bin\Debug\com.MyApp.MyApp-arm64-v8a-Signed.apk" 
Output file 'bin\Debug\com.MyApp.MyApp-arm64-v8a-Signed.apk' exists
"zipalign.exe" exited with code 1.
Done executing task "AndroidZipAlign" -- FAILED

### Basic Information

VS 15.6
XF 2.4.0.38779

I'm not the only one seeing this. @batmaci has also reported it at https://forums.xamarin.com/discussion/comment/325764


_Copied from original issue: xamarin/Xamarin.Forms#2243_

Is this what you mean 

`Build started 09/04/2018 13:53:10.
__________________________________________________
Project "/Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/testHelloworld2.csproj" (Install target(s)):

Target _SetLatestTargetFrameworkVersion:
    Found Java SDK version 1.8.0.
Target _ValidateAndroidPackageProperties:
      PackageName: com.companyname.testHelloworld2
Target _ResolveMonoAndroidSdks:
    MonoAndroid Tools: /Library/Frameworks/Xamarin.Android.framework/Libraries/xbuild/Xamarin/Android/
    Android Platform API level: 23
    TargetFrameworkVersion: v6.0
    Android NDK: /Users/myProjectFolder/Library/Developer/Xamarin/android-ndk/android-ndk-r14b/
    Android SDK: /Users/myProjectFolder/Library/Developer/Xamarin/android-sdk-macosx/
    Android SDK Build Tools: /Users/myProjectFolder/Library/Developer/Xamarin/android-sdk-macosx/build-tools/27.0.3/
    Java SDK: /usr/
Target _GetPrimaryCpuAbi:
    _PrimaryCpuAbi=arm64-v8a
Target _CheckInstantRunCondition:
    Dex Fast Deployment Enabled: False
Target _SetupInstantRun:
    Android Java Class for Fast Deployment: 
Target _SetupApplicationJavaClass:
    Application Java class: android.app.Application
Target _XamarinAndroidBuildResourceRestore:
    Reference has already had resources merged, skipping due to: obj/Debug/XbdMerge/Xamarin.Android.Support.Annotations.dll.stamp
Target _XamarinAndroidBuildAarRestore:
    Reference has already had resources merged, skipping due to: obj/Debug/XbdMerge/Xamarin.Android.Support.Compat.dll.stamp
    Reference has already had resources merged, skipping due to: obj/Debug/XbdMerge/Xamarin.Android.Support.Core.UI.dll.stamp
    Reference has already had resources merged, skipping due to: obj/Debug/XbdMerge/Xamarin.Android.Support.Core.Utils.dll.stamp
    Reference has already had resources merged, skipping due to: obj/Debug/XbdMerge/Xamarin.Android.Support.Media.Compat.dll.stamp
    Reference has already had resources merged, skipping due to: obj/Debug/XbdMerge/Xamarin.Android.Support.Fragment.dll.stamp
    Reference has already had resources merged, skipping due to: obj/Debug/XbdMerge/Xamarin.Android.Support.Vector.Drawable.dll.stamp
    Reference has already had resources merged, skipping due to: obj/Debug/XbdMerge/Xamarin.Android.Support.Animated.Vector.Drawable.dll.stamp
    Reference has already had resources merged, skipping due to: obj/Debug/XbdMerge/Xamarin.Android.Support.v7.AppCompat.dll.stamp
Target _ResolveLibraryProjectImports:
  Skipping target "_ResolveLibraryProjectImports" because all output files are up-to-date with respect to the input files.
Target _BuildLibraryImportsCache:
  Skipping target "_BuildLibraryImportsCache" because all output files are up-to-date with respect to the input files.
Target _BuildAdditionalResourcesCache:
  Skipping target "_BuildAdditionalResourcesCache" because all output files are up-to-date with respect to the input files.
Target _CreateAdditionalResourceCache:
  Skipping target "_CreateAdditionalResourceCache" because it has no outputs.
Target _GenerateAndroidResourceDir:
  Skipping target "_GenerateAndroidResourceDir" because all output files are up-to-date with respect to the input files.
Target _GenerateCodeBehindForLayout:
  Skipping target "_GenerateCodeBehindForLayout" because it has no outputs.
Target _DefineBuildTargetAbis:
    Build target ABI: armeabi-v7a;x86
Target VectorDrawableCheckBuildToolsVersionTask:
    Checking Android SDK Build-tools version...
    Selected Android SDK Build Tools Path: /Users/myProjectFolder/Library/Developer/Xamarin/android-sdk-macosx/build-tools/27.0.3/
    Selected Android SDK Build Tools Version: 27.0.3
    Android SDK Build Tools Version: 27.0.3 meets minimum requirements for Vector Drawables. OK.
    Finished Checking Android SDK Build-tools version.
Target _UpdateAndroidResgen:
  Skipping target "_UpdateAndroidResgen" because all output files are up-to-date with respect to the input files.
Target _CreateAdditionalResourceCache:
  Skipping target "_CreateAdditionalResourceCache" because it has no outputs.
Target _GenerateJavaDesignerForComponent:
  Skipping target "_GenerateJavaDesignerForComponent" because all output files are up-to-date with respect to the input files.
Target GenerateTargetFrameworkMonikerAttribute:
  Skipping target "GenerateTargetFrameworkMonikerAttribute" because all output files are up-to-date with respect to the input files.
Target CoreCompile:
  Skipping target "CoreCompile" because all output files are up-to-date with respect to the input files.
Target _CopyFilesMarkedCopyLocal:
    Touching "/Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/obj/Debug/testHelloworld2.csproj.CopyComplete".
Target CopyFilesToOutputDirectory:
    testHelloworld2 -> /Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/bin/Debug/testHelloworld2.dll
Target IncrementalClean:
    Deleting file "/Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/obj/Debug/linksrc/testHelloworld2.pdb".
    Deleting file "/Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/obj/Debug/linksrc/Mono.Android.pdb".
    Deleting file "/Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/bin/Debug/Mono.Android.pdb".
Target _CheckTargetFrameworks:
    /Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/testHelloworld2.csproj : warning XA0105: The $(TargetFrameworkVersion) for Xamarin.Android.Support.Animated.Vector.Drawable.dll (v7.0) is greater than the $(TargetFrameworkVersion) for your project (v6.0). You need to increase the $(TargetFrameworkVersion) for your project.
    /Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/testHelloworld2.csproj : warning XA0105: The $(TargetFrameworkVersion) for Xamarin.Android.Support.Annotations.dll (v7.0) is greater than the $(TargetFrameworkVersion) for your project (v6.0). You need to increase the $(TargetFrameworkVersion) for your project.
    /Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/testHelloworld2.csproj : warning XA0105: The $(TargetFrameworkVersion) for Xamarin.Android.Support.Compat.dll (v7.0) is greater than the $(TargetFrameworkVersion) for your project (v6.0). You need to increase the $(TargetFrameworkVersion) for your project.
    /Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/testHelloworld2.csproj : warning XA0105: The $(TargetFrameworkVersion) for Xamarin.Android.Support.Core.UI.dll (v7.0) is greater than the $(TargetFrameworkVersion) for your project (v6.0). You need to increase the $(TargetFrameworkVersion) for your project.
    /Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/testHelloworld2.csproj : warning XA0105: The $(TargetFrameworkVersion) for Xamarin.Android.Support.Core.Utils.dll (v7.0) is greater than the $(TargetFrameworkVersion) for your project (v6.0). You need to increase the $(TargetFrameworkVersion) for your project.
    /Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/testHelloworld2.csproj : warning XA0105: The $(TargetFrameworkVersion) for Xamarin.Android.Support.Fragment.dll (v7.0) is greater than the $(TargetFrameworkVersion) for your project (v6.0). You need to increase the $(TargetFrameworkVersion) for your project.
    /Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/testHelloworld2.csproj : warning XA0105: The $(TargetFrameworkVersion) for Xamarin.Android.Support.Media.Compat.dll (v7.0) is greater than the $(TargetFrameworkVersion) for your project (v6.0). You need to increase the $(TargetFrameworkVersion) for your project.
    /Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/testHelloworld2.csproj : warning XA0105: The $(TargetFrameworkVersion) for Xamarin.Android.Support.v7.AppCompat.dll (v7.0) is greater than the $(TargetFrameworkVersion) for your project (v6.0). You need to increase the $(TargetFrameworkVersion) for your project.
    /Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/testHelloworld2.csproj : warning XA0105: The $(TargetFrameworkVersion) for Xamarin.Android.Support.Vector.Drawable.dll (v7.0) is greater than the $(TargetFrameworkVersion) for your project (v6.0). You need to increase the $(TargetFrameworkVersion) for your project.
Done building target "_CheckTargetFrameworks" in project "testHelloworld2.csproj".
Target _CollectConfigFiles:
  Skipping target "_CollectConfigFiles" because all output files are up-to-date with respect to the input files.
Target _CopyConfigFiles:
  Skipping target "_CopyConfigFiles" because it has no outputs.
Target _CopyIntermediateAssemblies:
  Skipping target "_CopyIntermediateAssemblies" because all output files are up-to-date with respect to the input files.
Target _ConvertPdbFiles:
  Skipping target "_ConvertPdbFiles" because it has no outputs.
Target _CopyPdbFiles:
    Copying file from "bin/Debug/testHelloworld2.pdb" to "obj/Debug/linksrc/testHelloworld2.pdb".
    Copying file from "bin/Debug/Mono.Android.pdb" to "obj/Debug/linksrc/Mono.Android.pdb".
Target _CopyMdbFiles:
  Skipping target "_CopyMdbFiles" because it has no outputs.
Target _StripEmbeddedLibraries:
  Skipping target "_StripEmbeddedLibraries" because all output files are up-to-date with respect to the input files.
Target _LinkAssembliesNoShrink:
    Touching "obj/Debug/link.flag".
Target _AddStaticResources:
  Skipping target "_AddStaticResources" because all output files are up-to-date with respect to the input files.
Target _CreateAdditionalResourceCache:
  Skipping target "_CreateAdditionalResourceCache" because it has no outputs.
Target _GenerateAndroidAssetsDir:
  Skipping target "_GenerateAndroidAssetsDir" because it has no outputs.
Target _CreateBaseApk:
  Skipping target "_CreateBaseApk" because all output files are up-to-date with respect to the input files.
Target _CreateAdditionalResourceCache:
  Skipping target "_CreateAdditionalResourceCache" because it has no outputs.
Target _CompileJava:
  Skipping target "_CompileJava" because all output files are up-to-date with respect to the input files.
Target _CreateAdditionalResourceCache:
  Skipping target "_CreateAdditionalResourceCache" because it has no outputs.
Target _CompileToDalvikWithDx:
  Skipping target "_CompileToDalvikWithDx" because all output files are up-to-date with respect to the input files.
Target _CheckApkPerAbiFlag:
    Touching "obj/Debug/android/bin/apk_per_abi.flag".
Target _CopyPackage:
    Deleting file "bin/Debug/com.companyname.testHelloworld2.apk".
    Copying file from "/Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/obj/Debug/android/bin/com.companyname.testHelloworld2.apk" to "bin/Debug/com.companyname.testHelloworld2.apk".
    Copying file from "/Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/obj/Debug/android/bin/com.companyname.testHelloworld2-armeabi-v7a.apk" to "bin/Debug/com.companyname.testHelloworld2-armeabi-v7a.apk".
    Copying file from "/Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/obj/Debug/android/bin/com.companyname.testHelloworld2-x86.apk" to "bin/Debug/com.companyname.testHelloworld2-x86.apk".
Target _ResolveAndroidSigningKey:
    Touching "obj/Debug/android_debug_keystore.flag".
Target _Sign:
    /usr/bin/keytool -list -alias androiddebugkey -storepass android -keypass android -keystore "/Users/myProjectFolder/.local/share/Xamarin/Mono for Android/debug.keystore" 
    Deleting file "bin/Debug/com.companyname.testHelloworld2-Signed.apk".
    /Users/myProjectFolder/Library/Developer/Xamarin/android-sdk-macosx/build-tools/27.0.3/zipalign 4 "/Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/obj/Debug/android/bin/com.companyname.testHelloworld2.apk" "bin/Debug//com.companyname.testHelloworld2-Signed.apk" 
    /Users/myProjectFolder/Library/Developer/Xamarin/android-sdk-macosx/build-tools/27.0.3/zipalign 4 "/Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/obj/Debug/android/bin/com.companyname.testHelloworld2-armeabi-v7a.apk" "bin/Debug//com.companyname.testHelloworld2-armeabi-v7a-Signed.apk" 
    Output file 'bin/Debug//com.companyname.testHelloworld2-armeabi-v7a-Signed.apk' exists
    /Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.targets(2588,2): error MSB6006: "zipalign" exited with code 1.
Done building target "_Sign" in project "testHelloworld2.csproj" -- FAILED.

Done building project "testHelloworld2.csproj" -- FAILED.

Build FAILED.

/Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/testHelloworld2.csproj : warning XA0105: The $(TargetFrameworkVersion) for Xamarin.Android.Support.Animated.Vector.Drawable.dll (v7.0) is greater than the $(TargetFrameworkVersion) for your project (v6.0). You need to increase the $(TargetFrameworkVersion) for your project.
/Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/testHelloworld2.csproj : warning XA0105: The $(TargetFrameworkVersion) for Xamarin.Android.Support.Annotations.dll (v7.0) is greater than the $(TargetFrameworkVersion) for your project (v6.0). You need to increase the $(TargetFrameworkVersion) for your project.
/Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/testHelloworld2.csproj : warning XA0105: The $(TargetFrameworkVersion) for Xamarin.Android.Support.Compat.dll (v7.0) is greater than the $(TargetFrameworkVersion) for your project (v6.0). You need to increase the $(TargetFrameworkVersion) for your project.
/Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/testHelloworld2.csproj : warning XA0105: The $(TargetFrameworkVersion) for Xamarin.Android.Support.Core.UI.dll (v7.0) is greater than the $(TargetFrameworkVersion) for your project (v6.0). You need to increase the $(TargetFrameworkVersion) for your project.
/Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/testHelloworld2.csproj : warning XA0105: The $(TargetFrameworkVersion) for Xamarin.Android.Support.Core.Utils.dll (v7.0) is greater than the $(TargetFrameworkVersion) for your project (v6.0). You need to increase the $(TargetFrameworkVersion) for your project.
/Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/testHelloworld2.csproj : warning XA0105: The $(TargetFrameworkVersion) for Xamarin.Android.Support.Fragment.dll (v7.0) is greater than the $(TargetFrameworkVersion) for your project (v6.0). You need to increase the $(TargetFrameworkVersion) for your project.
/Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/testHelloworld2.csproj : warning XA0105: The $(TargetFrameworkVersion) for Xamarin.Android.Support.Media.Compat.dll (v7.0) is greater than the $(TargetFrameworkVersion) for your project (v6.0). You need to increase the $(TargetFrameworkVersion) for your project.
/Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/testHelloworld2.csproj : warning XA0105: The $(TargetFrameworkVersion) for Xamarin.Android.Support.v7.AppCompat.dll (v7.0) is greater than the $(TargetFrameworkVersion) for your project (v6.0). You need to increase the $(TargetFrameworkVersion) for your project.
/Users/myProjectFolder/Projects/testHelloworld2/testHelloworld2/testHelloworld2.csproj : warning XA0105: The $(TargetFrameworkVersion) for Xamarin.Android.Support.Vector.Drawable.dll (v7.0) is greater than the $(TargetFrameworkVersion) for your project (v6.0). You need to increase the $(TargetFrameworkVersion) for your project.
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.targets(2588,2): error MSB6006: "zipalign" exited with code 1.
    9 Warning(s)
    1 Error(s)

Time Elapsed 00:00:05.07`

--- a/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/PackagingTest.cs
+++ b/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/PackagingTest.cs
@@ -204,24 +204,29 @@ namespace Xamarin.Android.Build.Tests
 		}
 
 		[Test]
-		public void CheckSignApk ([Values(true, false)] bool useApkSigner)
+		public void CheckSignApk ([Values(true, false)] bool useApkSigner, [Values(true, false)] bool perAbiApk)
 		{
 			string ext = Environment.OSVersion.Platform != PlatformID.Unix ? ".exe" : "";
-			if (useApkSigner && !File.Exists (Path.Combine (AndroidSdkPath, "build-tools", "26.0.1", "apksigner"+ ext))) {
-				Assert.Ignore ("Skipping test. Required build-tools verison 26.0.1 is not installed.");
+			var foundApkSigner = Directory.EnumerateDirectories (Path.Combine (AndroidSdkPath, "build-tools")).Any (dir => Directory.EnumerateFiles (dir, "apksigner"+ ext).Any ());
+			if (useApkSigner && !foundApkSigner) {
+				Assert.Ignore ("Skipping test. Required build-tools verison which contains apksigner is not installed.");
 			}
 			var proj = new XamarinAndroidApplicationProject () {
 				IsRelease = true,
 			};
 			if (useApkSigner) {
 				proj.SetProperty ("AndroidUseApkSigner", "true");
-				proj.SetProperty ("AndroidSdkBuildToolsVersion", "26.0.1");
+	//			proj.SetProperty ("AndroidSdkBuildToolsVersion", "26.0.1");
 			} else {
 				proj.RemoveProperty ("AndroidUseApkSigner");
 			}
+			proj.SetProperty (proj.ReleaseProperties, KnownProperties.AndroidCreatePackagePerAbi, perAbiApk);
+			proj.SetProperty (proj.ReleaseProperties, KnownProperties.AndroidSupportedAbis, "armeabi-v7a;x86");
 			using (var b = CreateApkBuilder (Path.Combine ("temp", TestContext.CurrentContext.Test.Name))) {
 				b.Verbosity = Microsoft.Build.Framework.LoggerVerbosity.Diagnostic;
 				Assert.IsTrue (b.Build (proj), "build failed");
+				proj.AndroidResources.First ().Timestamp = null;
+				Assert.IsTrue (b.Build (proj), "Second build failed");
 			}
 		}
 

--- a/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/PackagingTest.cs
+++ b/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/PackagingTest.cs
@@ -216,7 +216,6 @@ namespace Xamarin.Android.Build.Tests
 			};
 			if (useApkSigner) {
 				proj.SetProperty ("AndroidUseApkSigner", "true");
-	//			proj.SetProperty ("AndroidSdkBuildToolsVersion", "26.0.1");
 			} else {
 				proj.RemoveProperty ("AndroidUseApkSigner");
 			}

--- a/src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets
+++ b/src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets
@@ -2612,7 +2612,11 @@ because xbuild doesn't support framework reference assemblies.
 		TimestampAuthorityCertificateAlias="$(JarsignerTimestampAuthorityCertificateAlias)"
 		SigningAlgorithm="$(AndroidApkSigningAlgorithm)"
 	/>
-	<Delete Files="$(ApkFileSigned)" Condition=" '$(AndroidUseApkSigner)' == 'true' "/>
+	<ItemGroup>
+		<ApkAbiFilesSigned Include="$(ApkFileSigned)" Condition="'$(AndroidUseApkSigner)' == 'true'" />
+		<ApkAbiFilesSigned Condition="'$(AndroidCreatePackagePerAbi)' == 'true' And '$(AndroidUseApkSigner)' == 'true' " Include="$(OutDir)$(_AndroidPackage)*-Signed.apk" />
+	</ItemGroup>
+	<Delete Files="%(ApkAbiFilesSigned.FullPath)" Condition=" '$(AndroidUseApkSigner)' == 'true' "/>
 	<AndroidZipAlign Condition=" '$(AndroidUseApkSigner)' == 'true' "
 		Source="%(ApkAbiFilesIntermediate.FullPath)"
 		DestinationDirectory="$(OutDir)"
@@ -2633,8 +2637,8 @@ because xbuild doesn't support framework reference assemblies.
 	/>
 	<Message Text="Signed android package '$(ApkFileSigned)'" />
 	<ItemGroup>
-		<ApkAbiFilesSigned Include="$(ApkFileSigned)" />
-		<ApkAbiFilesSigned Condition="'$(AndroidCreatePackagePerAbi)' == 'true'" Include="$(OutDir)$(_AndroidPackage)*-Signed.apk" />
+		<ApkAbiFilesSigned Include="$(ApkFileSigned)" Condition="'$(AndroidUseApkSigner)' != 'true'" />
+		<ApkAbiFilesSigned Condition="'$(AndroidCreatePackagePerAbi)' == 'true' And '$(AndroidUseApkSigner)' != 'true'" Include="$(OutDir)$(_AndroidPackage)*-Signed.apk" />
 	</ItemGroup>
 	<Delete Files="%(ApkAbiFilesSigned.FullPath)" Condition=" '$(AndroidUseApkSigner)' != 'true' "/>
 	<ItemGroup>

